includes/razorpay-webhook.php

```diff
@@ -116,4 +150,134 @@ protected function paymentAuthorized($data)
+        //
+        $orderId = $data['payload']['payment']['entity']['notes']['woocommerce_order_id'];
+
+        $paymentId = $data['payload']['payment']['entity']['id'];
+
+        if (isset($data['payload']['payment']['subscription_id']) === true)
+        {
+            $this->processSubscription($orderId, $paymentId, false);
+        }
+
+        exit;
+    }
+
+    /**
+     * Handling the subscription charged webhook
+     *
+     * @param $data
+     */
+    protected function subscriptionCharged($data)
+    {
+        //
+        // Order entity should be sent as part of the webhook payload
+        //
+        $orderId = $data['payload']['subscription']['entity']['notes']['woocommerce_order_id'];
+
+        $this->processSubscription($orderId);
+
+        exit;
+    }
+
+    /**
+     * Helper method used to handle all subscription processing
+     *
+     * @param $orderId
+     * @param $paymentId
+     * @param $success
+     */
+    protected function processSubscription($orderId, $paymentId, $success = true)
+    {
+        //
+        // If success is false, automatically process subscription failure
+        //
+        if ($success === false)
+        {
+            return $this->processSubscriptionFailed($orderId);
+        }
+
+        // This method is used to process the subscription's recurring payment
+        $wcSubscription = wcs_get_subscriptions_for_order($orderId);
+
+        $subscriptionId = get_post_meta($orderId, self::RAZORPAY_SUBSCRIPTION_ID)[0];
+
+        $api = $this->razorpay->getRazorpayApiInstance();
+
+        try
+        {
+            $subscription = $api->subscription->fetch($subscriptionId);
+        }
+        catch (Exception $e)
+        {
+            echo "Razorpay Error: " . $e->getMessage();
+        }
+
+        $this->processSubscriptionSuccess($wcSubscription, $subscription, $paymentId);
+
+        exit;
+    }
+
+    /**
+     * In the case of successful payment, we mark the subscription successful
+     *
+     * @param $wcSubscription
+     * @param $subscription
+     */
+    protected function processSubscriptionSuccess($wcSubscription, $subscription, $paymentId)
+    {
+        if (count(array_keys($wcSubscription)) > 1)
+        {
+            $log = array(
+                'Error' => 'Array length of subscription array is more than 1'
+            );
+
+            write_log($log);
+
+            exit;
+        }
+
+        $wcSubscriptionId = array_keys($wcSubscription)[0];
+
+        // We will only process one subscription per order
+        $wcSubscription = array_values($wcSubscription)[0];
+
+        $paymentCount = $wcSubscription->get_completed_payment_count();
+
+        // The subscription is completely paid for
+        if ($paymentCount === $subscription->total_count)
+        {
+            return;
+        }
+        else if ($wcSubscription->get_completed_payment_count() + 1 === $subscription->paid_count)
```

Comment:
@StSarc I prefer having this check on here. Although, it may not be required. @captn3m0 could you add your point of view here?